Update codes for workitem #149 - More features on spell checking

WhiteHsu opened the pull request:
@sebbrudzinski 

I've updated codes for workitem #149 - More features on spell checking. Here's some more information about these updates:
1. For now there's another thread for automatic spell check. Once you open a file or save any modifications, spell check will be triggered automatically:
![default](https://cloud.githubusercontent.com/assets/7831418/26093102/f6903692-3a46-11e7-8df3-1e076a9aa78c.png)

2. Because there's another thread, for now it'll not "hanged" will running spell check.

3. If you modify anything without saving changes, because Editor State isn't dirty, spell check will not be triggered:
![default](https://cloud.githubusercontent.com/assets/7831418/26093140/328e070a-3a47-11e7-9250-452465f485ea.png)

4. However, you can click the Icon above to manually trigger spell check, then i'll be executed still without saving changes:
![default](https://cloud.githubusercontent.com/assets/7831418/26093176/552d6e18-3a47-11e7-93be-7dac6a6b6e63.png)

5. Please help review and feedback. Thank you very much.  :-)

WhiteHsu commented:
@sebbrudzinski 
Please kindly help review and suggest on my code updates. Thank you very much.  :-)

sebbrudzinski commented on OpenLaTeXStudio/Editor/src/main/java/latexstudio/editor/toolbar/SpellCheckService.java:
```diff
@@ -0,0 +1,46 @@
+            if (etc.getEditorState().isDirty()) {
+                try {
+                    etc.getEditorState().setDirty(false);
+                    etc.spellCheckAllText();
```
Would it be possible to move the whole "spellCheckAllText" method to this class? I don't think it's still used anywhere else.

WhiteHsu commented:
@sebbrudzinski 
Thanks for your comments. I've already moved the method as your suggestion but encountered issue to "disable spell checking after it has been enabled".

I guess what you mean is: We should make the "Spell Check" button different status in the Toolbar after enabling this function. Therefore, for now after enabling spell check, the button will be changed on Toolbar:
![default](https://user-images.githubusercontent.com/7831418/27092770-0b9c5352-5097-11e7-9582-93078316cba1.png)

And if you click again, Spell Check function will be disabled, and the "Spell Check" button status will be restored to the original status in Toolbar:
![default](https://user-images.githubusercontent.com/7831418/27092791-246ca652-5097-11e7-8cef-d3c068064d4a.png)

I've implemented this change by "setBorder()". I call "setBorder()" method once getting the triggering source with "e.getSource()" in the actionPerformed method:
![default](https://user-images.githubusercontent.com/7831418/27092854-5cc25b82-5097-11e7-80c2-45505d69c356.png)
![default](https://user-images.githubusercontent.com/7831418/27092862-61918cc8-5097-11e7-8594-20f4559d65f0.png)

However, the encountered problem is: If we click the menu item, this method will throw exception:
![default](https://user-images.githubusercontent.com/7831418/27092977-bae7448e-5097-11e7-90a8-c18cceb99099.png)
![default](https://user-images.githubusercontent.com/7831418/27092984-bec409ca-5097-11e7-844d-4a427dd37e9d.png)

The exception reason is: If you click menu item instead of Toolbar Button, the "e.getSource()" in the actionPerformed method will not get the Toolbar Button Object but the Menu Item Object, then we cannot use "e.getSource()" to change the Spell Check button status.
![default](https://user-images.githubusercontent.com/7831418/27093475-6b4a9f5a-5099-11e7-9189-0aa9d0d25e58.png)

We may be able to get the Spell Check button to doClick() or trigger the actionPerformed() manually, but after trying in these days, I cannot implement both of them. I've ever tried:
1. Try to get the Spell Check button by ID, but it seems the way in Javascript but cannot work among Java Swing?
2. Try to get sub-components from the root pane but eventually cannot get this Spell Check button object after digging down with several levels.
3. Try to call the actionPerformed() method directly, but I dont know how to assign the triggering source as the Spell Check button?
4. Try to get the Toolbar directly but cannot find the way to do it, including SwingUtilities and WindowManager.

So far I do not have other ideas and would like to ask your advice. Or should we remove this item from Menu and just keep the button in Toolbar? Please kindly feedback and suggest. Thank you very much.

WhiteHsu commented:
@sebbrudzinski 
I've just found the way to get Toobar. So I'll update the codes and submit request again soon. Many thanks.  :-)

WhiteHsu commented:
@sebbrudzinski 
I have further updates based on your suggestions. Here're more details:
- To move the method "spellCheckAllText()" from ETC into SpellCheckService.java as local method.
- For now after enabling Spell Check, the Toolbar Icon will be changed to show that this function is enabled:
![default](https://user-images.githubusercontent.com/7831418/27221740-47b5790a-52bc-11e7-88c2-b2a9fd0b5d92.png)
- The Menu Item in the Menubar will be also changed simultaneously:
![default](https://user-images.githubusercontent.com/7831418/27221807-73c33a28-52bc-11e7-9e94-a63a2c0ad9bb.png)
- If we disable Spell Check, these two items will be resumed to its original UI status:
![default](https://user-images.githubusercontent.com/7831418/27221848-9d3e91a4-52bc-11e7-9e2a-dac6aa79c8dd.png)
![default](https://user-images.githubusercontent.com/7831418/27221853-a23321b6-52bc-11e7-96fb-1cec4e81fd0f.png)
- These changes can be triggered from both Menubar Item and Toolbar Icon. I don't really "disable spell checking after it has been enabled", because once disabling them, then users will never be able to click the item/icon again. So I only make them UI changes while enabling or disabling occurs.

Please kindly review and suggest. Thank you very much.  :-)

WhiteHsu commented:
@sebbrudzinski 
Thanks for your helpful feedback. I've updated codes based on your suggestion. Please help review. Thank you.  :-)

WhiteHsu commented:
Got it. Many thanks!!  :-)
